### ISSUE TITLE
MAC M1 GPU Support

### ISSUE THREAD

**Issue by santoshbhor** [xmrig/xmrig#2706]
Hi I wanted to check and see if GPU support from MAC M1 can be added and what would take to get this done.
I can possibly help but i am not MAC M1 expert but if some one can guide me i may be able to help to support this?

**Comment by Spudz76** [comment]
Already works, OpenCL.

Not that fast since Apple OpenCL is terrible (for example identical AMD cards about 60% performance of AMD OpenCL on Linux).  Won't really work simultaneous with CPU mining since they share lots of the total chip bandwidth (memory bus mostly).

But it definitely should just work in recent versions by simply enabling OpenCL backend.

**Comment by cstrouse** [comment]
I tried out GPU support just for fun on my 8/8 16GB MacBookAir10,1 running macOS 12.0.1 and it doesn't seem to work due to the OpenCL build failing.

Here is the relevant bit of the log:
```
[2021-11-22 21:54:28.255]  opencl   use profile  rx  (1 thread) scratchpad 2048 KB
|  # | GPU |  BUS ID | INTENSITY | WSIZE | MEMORY | NAME
|  0 |   0 |     n/a |       128 |     8 |    256 | Apple M1
[2021-11-22 21:54:28.263]  opencl   GPU #0 compiling...
UNSUPPORTED (log once): buildComputeProgram: cl2Metal failed
[2021-11-22 21:54:28.666]  opencl   error CL_BUILD_PROGRAM_FAILURE when calling clBuildProgram
BUILD LOG:
Compilation failed:

program_source:1788:1: warning: comparison of integers of different signs: 'uint32_t' (aka 'unsigned int') and 'int'
update_max(latency,(last_memory_op_slot+WORKERS_PER_HASH)/WORKERS_PER_HASH);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:1399:56: note: expanded from macro 'update_max'
#define update_max(value, next_value) do { if ((value) < (next_value)) (value) = (next_value); } while (0)
                                                ~~~~~  ^  ~~~~~~~~~~
program_source:1811:1: warning: comparison of integers of different signs: 'int32_t' (aka 'int') and 'unsigned int'
update_max(first_allowed_slot,latency*WORKERS_PER_HASH);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:1399:56: note: expanded from macro 'update_max'
#define update_max(value, next_value) do { if ((value) < (next_value)) (value) = (next_value); } while (0)
                                                ~~~~~  ^  ~~~~~~~~~~
[2021-11-22 21:54:28.667]  opencl   thread #0 failed with error CL_INVALID_PROGRAM
[2021-11-22 21:54:28.668]  opencl   thread #0 self-test failed
[2021-11-22 21:54:28.668]  opencl   disabled (failed to start threads)
[2021-11-22 21:54:36.434]  signal   Ctrl+C received, exiting
[2021-11-22 21:54:36.437]  cpu      stopped (2 ms)
[2021-11-22 21:54:36.437]  opencl   stopped (0 ms)
```

I know M1 support is not likely a priority since there's not much benefit to be had and probably doesn't make much sense to use for this purpose but hope the log helps someone who is interested in improving support.

**Comment by Spudz76** [comment]
Well, it used to work.  Not sure what Apple changed.

**Comment by cstrouse** [comment]
Sorry for the noise guys, it does start with some additional arguments it turns out. Try the following and see if you get it to build the OpenCL payload and start.

`./xmrig --coin=RVN --no-cpu --opencl --opencl --opencl-platform=0 -o stratum+tcp://stratum.ravenpool.ninja:3333 -u RVNWALLET.WORKERNAME -p x`

I'm able to get 1.76 MH/s using kawpow on 12.0.1 using a Radeon Pro WX 5100 8GB card and the M1 MacBookAir10,1 (8-core CPU, 8-core GPU, 16GB of memory) gets 1.41 MH/s using kawpow. I ran the miner for over an hour on both machines and neither ever submitted any shares to the pool so don't recommend it.

**Comment by Spudz76** [comment]
Yes, unmineable has a minimum diff that is too high for "slow" rigs.

I am not sure if RandomX ever worked, besides that it would perform worse than horrible.  Just noticed that other attempt was rx.

**Comment by geoffwade** [comment]
> This worked great on my macbook Air M1.   Seeing about 1.38 MH/s.  Not a lot of info out there for M1 GPU mining and happy to see this worked.  Thank you!



> Sorry for the noise guys, it does start with some additional arguments it turns out. Try the following and see if you get it to build the OpenCL payload and start.
> 
> `./xmrig --coin=RVN --no-cpu --opencl --opencl-platform=0 -o stratum+tcp://stratum.ravenpool.ninja:3333 -u RVNWALLET.WORKERNAME -p x`
> 
> I'm able to get 1.76 MH/s using kawpow on 12.0.1 using a Radeon Pro WX 5100 8GB card and the M1 MacBookAir10,1 (8-core CPU, 8-core GPU, 16GB of memory) gets 1.41 MH/s using kawpow. I ran the miner for over an hour on both machines and neither ever submitted any shares to the pool so don't recommend it (on unmineable which I tried first; other pools do seem to work properly).

**Comment by acrogenesis** [comment]
> Sorry for the noise guys, it does start with some additional arguments it turns out. Try the following and see if you get it to build the OpenCL payload and start.
> 
> `./xmrig --coin=RVN --no-cpu --opencl --opencl-platform=0 -o stratum+tcp://stratum.ravenpool.ninja:3333 -u RVNWALLET.WORKERNAME -p x`
> 
> I'm able to get 1.76 MH/s using kawpow on 12.0.1 using a Radeon Pro WX 5100 8GB card and the M1 MacBookAir10,1 (8-core CPU, 8-core GPU, 16GB of memory) gets 1.41 MH/s using kawpow. I ran the miner for over an hour on both machines and neither ever submitted any shares to the pool so don't recommend it (on unmineable which I tried first; other pools do seem to work properly).

I'm getting 6.21 MH/s on the M1 max 16" 32GB. Although I do have a lot of windows and apps open

**Comment by Spudz76** [comment]
Pretty good considering the watts are low, and doing other stuff.

**Comment by RetroAndDev** [comment]
Any help on that ? 
I active the OpenCL backend on M1 and I got this error : 
 * ABOUT        XMRig/6.21.0 clang/13.0.0 (built for macOS x86-64, 64 bit)
 * LIBS         libuv/1.44.2 OpenSSL/3.0.7 hwloc/2.9.0
 * HUGE PAGES   supported
 * 1GB PAGES    unavailable
 * CPU          VirtualApple @ 2.50GHz (1) 64-bit AES
                L2:16.0 MB L3:0.0 MB 8C/8T NUMA:1
 * MEMORY       7.9/8.0 GB (99%)
 * DONATE       1%
 * ASSEMBLY     auto:intel
 * POOL #1      rx.unmineable.com:3333 algo rx/0
 * COMMANDS     hashrate, pause, resume, results, connection
 * OPENCL       #0 Apple/OpenCL 1.2 (Nov 11 2023 23:48:03)
 * OPENCL GPU   #0 n/a Apple M1 1000 MHz cu:8 mem:1024/5461 MB
 * CUDA         disabled
[2024-01-12 06:48:30.559]  net      use pool rx.unmineable.com:3333  161.35.34.195
[2024-01-12 06:48:30.560]  net      new job from rx.unmineable.com:3333 diff 100001 algo rx/0 height 162169
[2024-01-12 06:48:30.560]  cpu      use argon2 implementation SSSE3
[2024-01-12 06:48:30.560]  randomx  init dataset algo rx/0 (8 threads) seed 8d47b82e037755c8...
[2024-01-12 06:48:30.561]  randomx  allocated 2336 MB (2080+256) huge pages 0% 0/1168 +JIT (0 ms)
[2024-01-12 06:48:37.775]  randomx  dataset ready (7214 ms)
[2024-01-12 06:48:37.796]  opencl   use profile  rx  (1 thread) scratchpad 2048 KB
|  # | GPU |  BUS ID | INTENSITY | WSIZE | MEMORY | NAME
|  0 |   0 |     n/a |       128 |     8 |    256 | Apple M1
[2024-01-12 06:48:37.798]  opencl   GPU #0 compiling...
UNSUPPORTED (log once): buildComputeProgram: cl2Metal failed
[2024-01-12 06:48:38.160]  opencl   error CL_BUILD_PROGRAM_FAILURE when calling clBuildProgram
BUILD LOG:
program_source:1786:1: warning: comparison of integers of different signs: '__private uint32_t' (aka '__private unsigned int') and 'int' [-Wsign-compare]
update_max(latency,(last_memory_op_slot+WORKERS_PER_HASH)/WORKERS_PER_HASH);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:1397:56: note: expanded from macro 'update_max'
#define update_max(value, next_value) do { if ((value) < (next_value)) (value) = (next_value); } while (0)
                                                ~~~~~  ^  ~~~~~~~~~~
program_source:1809:1: warning: comparison of integers of different signs: '__private int32_t' (aka '__private int') and 'unsigned int' [-Wsign-compare]
update_max(first_allowed_slot,latency*WORKERS_PER_HASH);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
program_source:1397:56: note: expanded from macro 'update_max'
#define update_max(value, next_value) do { if ((value) < (next_value)) (value) = (next_value); } while (0)
                                                ~~~~~  ^  ~~~~~~~~~~
[2024-01-12 06:48:38.160]  opencl   thread #0 failed with error CL_INVALID_PROGRAM
[2024-01-12 06:48:38.161]  opencl   thread #0 self-test failed
[2024-01-12 06:48:38.161]  opencl   disabled (failed to start threads)

Any idea ? I used the Intel x86_64 and ARM built and I got the same error.

**Comment by SChernykh** [comment]
OpenCL on MacOS is not supported, and you shouldn't be mining RandomX with a GPU at all.

**Comment by RetroAndDev** [comment]
Ah ! 
It's possible to mine on MacOS with M1 on other miners ? To make profitable my MacBook ?

**Comment by RetroAndDev** [comment]
Yeah, but performances are low ? I have 2kH/s and with ETHMiner-M1 Fork is a 2MH/s. That good ?
(I can't use ETHMiner in my pool, don't know why)

**Comment by RetroAndDev** [comment]
Oh okay. But there is no ways to get the full power of the M1 (CPU+GPU) ?

**Comment by SChernykh** [comment]
You can try to run XMRig on CPU and ETHMiner on GPU, but they will be interfering with each other because both use memory bandwidth for mining.

**Comment by RetroAndDev** [comment]
ETHMiner don't let me mine with unMinable and ETC. Too had

**Comment by Spudz76** [comment]
Seems like [vk-ethminer](https://github.com/mshzhb/vk-ethminer) might work for ETC on M1 GPUs (using Vulkan instead of buggy/abandoned Apple OpenCL).

**Comment by RetroAndDev** [comment]
Sure !
That's sounds good 👌 
I will try it. Sure that Vulcan is better than OpenCL on Macs

**Comment by RetroAndDev** [comment]
> Seems like [vk-ethminer](https://github.com/mshzhb/vk-ethminer) might work for ETC on M1 GPUs (using Vulkan instead of buggy/abandoned Apple OpenCL).

Got error : 
Fatal : VkResult is -9 in /Users/mshzhb/workspace/vulkan-ethminer/src/miner/vulkanminer.cpp at line 321
ERROR:             vkEnumeratePhysicalDevices: Invalid instance [VUID-vkEnumeratePhysicalDevices-instance-parameter]
zsh: abort      ./vulkan_ethminer --server us1-etc.ethermine.org --port 4444 --wallet  --rig 

Someone know a fix ?
A Vulkan miner for low-power GPU is a great idea. Why there are no (good/big) miner(s) project(s) for passive mining ?

**Comment by Spudz76** [comment]
Might need to also install [MoltenVK](https://github.com/KhronosGroup/MoltenVK) layer so that Vulkan can get runtime translated to native Metal calls.  Maybe by default Apple has their own also-broken implementation of Vulkan.  They prefer that everything but Metal is broken I think, to increase adoption/conversion.

**Comment by RetroAndDev** [comment]
Yeah but how ? I can't determine where is the issue cause of the source code is not available. 
I opened an issue on the project. Well see if I get answers

**Comment by ForbiddenEra** [comment]
> Sorry for the noise guys, it does start with some additional arguments it turns out. Try the following and see if you get it to build the OpenCL payload and start.
> 
> `./xmrig --coin=RVN --no-cpu --opencl --opencl-platform=0 -o stratum+tcp://stratum.ravenpool.ninja:3333 -u RVNWALLET.WORKERNAME -p x`
> 
> I'm able to get 1.76 MH/s using kawpow on 12.0.1 using a Radeon Pro WX 5100 8GB card and the M1 MacBookAir10,1 (8-core CPU, 8-core GPU, 16GB of memory) gets 1.41 MH/s using kawpow. I ran the miner for over an hour on both machines and neither ever submitted any shares to the pool so don't recommend it (on unmineable which I tried first; other pools do seem to work properly).

Nice.. I couldn't get it working for KAWPOW with the settings I tried in the config.json but this worked. I tried 'MacMiner' which seemed to work but I was getting zero accepted shares on NH.

Doing both randomx[6kH/s] and KAWPOW[8.5mH/s] on M1 Ultra Mac Studio; dual mining CPU+GPU doesn't seem to affect hashrate or either. For comparison though, I'm getting 25mH/s for KAWPOW on both my 3060+2070S. Also for comparison, I get about the same 6kH/s w/32 threads on my 4s*8c/16t (32c/64t) Xeon E4640 0 (albeit running virtualized)

**Comment by guyfischman** [comment]
Following up on this request — I've built a RandomX implementation on Apple Silicon via Metal. Current result on M4 Pro (20-core GPU): 515 H/s at 16W measured, 32.2 H/s/W. For context that's ~4× the best consumer GPU power efficiency number in sech1's RandomX_CUDA benchmark table (GTX 1660 Ti low-power at 8.1 H/s/W).

Byte-exact vs the reference RandomX implementation on the standard test vectors.

Absolute hashrate is still well below (I expect it will end up somewhere 10-20%) CPU xmrig on the same machine, so the use case is either running alongside CPU mining to recover otherwise-idle GPU power, or a reusable Metal RandomX primitive for verification/research tools.

@SChernykh Would you be open to accepting a Metal backend PR in principle?

**Comment by SChernykh** [comment]
Good question. Does it give better total hashrate when you mine both on CPU and GPU on your M4 Pro? Then it makes sense for people who don't care about power efficiency.

**Comment by guyfischman** [comment]
Good that you asked, last I tested that was when I was at 50H/s and it was all incremental. Right now when I run GPU full-power total hashrate (GPU+CPU) is 30% lower than CPU-only, I believe a little for SoC power throttling and mostly because they're contending for SLC. If I throttle GPU I can get +100-200H/s incrementally GPU+CPU over CPU.

So the answer is "yes provided it increases total hashrate"?

**Comment by SChernykh** [comment]
Huh, interesting. That SoC must be power limited, so the fact that it achieved higher hashrate with the same power budget is interesting. Yes, I guess it makes sense to add that code. Did you implement RandomX v2 as well though?

**Comment by guyfischman** [comment]
SoC power limit - my SoC uses 29W on CPU-only (with no core limit) and 33-34W for both, so I guess the budget is higher for SoC than CPU-block.

No v2 yet, I wanted to first see if a custom kernel could beat prior art first. Is v2 design final?

**Comment by SChernykh** [comment]
Yes, v2 is already released https://github.com/tevador/RandomX/releases/tag/v2.0
